### PR TITLE
all: ensure traces and stats for all receivers+exporters

### PR DESCRIPTION
**Pull request by odeke-em**
Ensure that every exporter and receiver has traces and stats.
Stats weren't available for any span recordings.

This change introduces a measurer that has an underlying span
and also tracks latency but also tags errors and statuses.

### CODE REVIEW

File: internal/observability.go [hunk 1]
```diff
@@ -65,42 +67,111 @@ var ViewExportedSpans = &view.View{
...
+// of activeness of the caller.
+func NewReceiverEventRecorder(receiverName string) SpansObservabilityRecorder {
+	return &spansRecorder{
+		predefinedSpanName: "opencensus.service.receiver." + receiverName + ".Receive",
+		startTime:          time.Now(),
+		receiverName:       receiverName,
+	}
+}
+
+// SpansObservabilityRecorder is a helper that is started when
```
Comment (bogdandrutu) [inline internal/observability.go]: incomplete comment.

File: internal/observability.go [hunk 2]
```diff
@@ -65,42 +67,111 @@ var ViewExportedSpans = &view.View{
...
+type SpansObservabilityRecorder interface {
```
Comment (bogdandrutu) [inline internal/observability.go]: why is this an interface instead of a struct?

File: internal/observability.go [hunk 3]
```diff
@@ -65,42 +67,111 @@ var ViewExportedSpans = &view.View{
...
-func ContextWithReceiverName(ctx context.Context, receiverName string) context.Context {
-	ctx, _ = tag.New(ctx, tag.Upsert(tagKeyReceiverName, receiverName))
+var _ SpansObservabilityRecorder = (*spansRecorder)(nil)
+
+func (sr *spansRecorder) Start(ctx context.Context, ni *commonpb.Node) context.Context {
+	sr.startTime = time.Now()
+	ctx, sr.span = trace.StartSpan(ctx, sr.predefinedSpanName, trace.WithSampler(trace.NeverSample()))
+
+	if sr.receiverName != "" {
+		ctx, _ = tag.New(ctx, tag.Upsert(tagKeyReceiverName, sr.receiverName))
```
Comment (bogdandrutu) [inline internal/observability.go]: Unnecessary overhead on the critical path the result of the`tag.Upsert(tagKeyReceiverName, sr.receiverName)` can be cached and reused.

Same everywhere.

File: internal/observability.go [hunk 4]
```diff
@@ -65,42 +67,111 @@ var ViewExportedSpans = &view.View{
...
 }
 
-// ContextWithReceiverName adds the tag "opencensus_receiver" and the name of the
-// receiver as the value, and returns the newly created context.
-func ContextWithReceiverName(ctx context.Context, receiverName string) context.Context {
-	ctx, _ = tag.New(ctx, tag.Upsert(tagKeyReceiverName, receiverName))
+var _ SpansObservabilityRecorder = (*spansRecorder)(nil)
+
+func (sr *spansRecorder) Start(ctx context.Context, ni *commonpb.Node) context.Context {
+	sr.startTime = time.Now()
```
Comment (bogdandrutu) [inline internal/observability.go]: Isn't this duplicated with line 173?

File: internal/observability.go [hunk 5]
```diff
@@ -65,42 +67,111 @@ var ViewExportedSpans = &view.View{
...
+		measurements = append(measurements, mReceivedSpans.M(nItems))
+	} else {
+		tagMutators = append(tagMutators, tag.Upsert(tagKeyExporterName, sr.exporterName))
+		measurements = append(measurements, mExportedSpans.M(nItems))
+	}
+
+	if err == nil {
+		tagMutators = append(tagMutators, tag.Upsert(tagKeyStatus, "OK"))
+	} else {
+		tagMutators = append(tagMutators, tag.Upsert(tagKeyStatus, "ERROR"), tag.Upsert(tagKeyError, err.Error()))
```
Comment (bogdandrutu) [inline internal/observability.go]: error is an int without any bounds, I don't think it is fine to put it as a tag in this case, that is the main reason we defined a type of errors (see the status in trace).

I would suggest to use if possible the same mapping that we have in trace (see https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto)

File: exporter/exporterparser/exparser.go [hunk 6]
```diff
@@ -43,28 +42,19 @@ import (
...
+	observabilityRecorder := internal.NewExporterEventRecorder(exporterName)
 
-	defer func() {
-		if aerr != nil {
-			span.SetStatus(trace.Status{Code: trace.StatusCodeInternal, Message: aerr.Error()})
-		}
-		span.End()
-	}()
+	ctx = observabilityRecorder.Start(ctx, td.Node)
+	defer observabilityRecorder.End(ctx, nGoodSpans, aerr)
```
Comment (pjanotti) [inline exporter/exporterparser/exparser.go]: nGoodSpans is going to be always zero when deferred function is called, either it becomes a capture or is passed as a reference.

File: exporter/zipkinexporter/zipkin.go [hunk 7]
```diff
@@ -193,18 +192,13 @@ func (ze *zipkinExporter) stop() error {
...
-	defer func() {
-		if zerr != nil {
-			span.SetStatus(trace.Status{Code: trace.StatusCodeInternal, Message: zerr.Error()})
-		}
-		span.End()
-	}()
+	observabilityRecorder := internal.NewExporterEventRecorder("zipkin")
+
+	ctx = observabilityRecorder.Start(ctx, td.Node)
+	defer observabilityRecorder.End(ctx, nGoodSpans, zerr)
```
Comment (pjanotti) [inline exporter/zipkinexporter/zipkin.go]: same as above

File: internal/observability.go [hunk 8]
```diff
@@ -65,42 +67,111 @@ var ViewExportedSpans = &view.View{
 	TagKeys:     []tag.Key{tagKeyExporterName},
 }
 
+var mLatencyMs = stats.Float64("oc.io/latency", "The latency of the various actions in milliseconds", "ms")
+
+var latencyDistribution = view.Distribution(
+	0, 0.01, 0.05, 0.1, 0.3, 0.6, 0.8, 1, 2, 3, 4, 5, 6, 8, 10, 13, 16, 20, 25, 30, 40, 50, 65, 80,
```
Comment (pjanotti) [inline internal/observability.go]: less than 0 is never going to be used, right? at least that is the way that this ends up in Prometheus.

File: internal/observability.go [hunk 9]
```diff
@@ -30,12 +31,13 @@ import (
 	"go.opencensus.io/trace"
 
 	commonpb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
-	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
 )
 
 var (
 	tagKeyReceiverName, _ = tag.NewKey("opencensus_receiver")
 	tagKeyExporterName, _ = tag.NewKey("opencensus_exporter")
+	tagKeyError, _        = tag.NewKey("opencensus_error")
```
Comment (pjanotti) [inline internal/observability.go]: The prefix "opencensus_" seems a bit to verbose to me: the namespace already should give that this is OpenCensus and then the paris will look strange: opencensus_receiver="zipkin". The case that makes sense is if for instance status is being mapped to the OC status defined in proto.

File: exporter/exporterparser/exparser.go (hunk 6 above)
Comment (pjanotti) [inline exporter/exporterparser/exparser.go]: deferred function is always going to be invoked with nGoodSpans set to zero, it either needs to be a capture or passed as reference.

File: exporter/opencensusexporter/opencensus.go [hunk 10]
```diff
@@ -105,20 +105,19 @@ func OpenCensusTraceExportersFromViper(v *viper.Viper) (tes []exporter.TraceExpo
 	return tes, mes, doneFns, nil
 }
 
-func (oce *ocagentExporter) ExportSpans(ctx context.Context, td data.TraceData) error {
+func (oce *ocagentExporter) ExportSpans(ctx context.Context, td data.TraceData) (err error) {
+	observabilityRecorder := internal.NewExporterEventRecorder("ocagent")
+
+	ctx = observabilityRecorder.Start(ctx, td.Node)
+	defer observabilityRecorder.End(ctx, int64(len(td.Spans)), err)
```
Comment (sjkaris) [inline exporter/opencensusexporter/opencensus.go]: err here will always be nil I believe. https://play.golang.org/p/kFOauu3Y7YF Similar to above, probably need a closure here

File: exporter/exporterparser/exparser.go (hunk 6 above)
Comment (sjkaris) [inline exporter/exporterparser/exparser.go]: `aerr` is never set

File: exporter/zipkinexporter/zipkin.go (hunk 7 above)
Comment (sjkaris) [inline exporter/zipkinexporter/zipkin.go]: `zerr` is never set. Also, as a nit, would prefer just `err` as a name here

File: internal/observability.go [hunk 11]
```diff
@@ -65,42 +67,111 @@ var ViewExportedSpans = &view.View{
...
+	Start(ctx context.Context, ni *commonpb.Node) context.Context
+	End(ctx context.Context, nItems int64, err error) context.Context
+	UnderlyingSpan() *trace.Span
+}
+
+type spansRecorder struct {
+	predefinedSpanName string
+	span               *trace.Span
+	receiverName       string
+	exporterName       string
```
Comment (sjkaris) [inline internal/observability.go]: I think you can form the wanted tags on construction (so they don't need to be reconstructed frequently), and then get rid of any receiver/exporter specificity within this struct itself

File: receiver/jaegerreceiver/trace_receiver.go [hunk 12]
```diff
@@ -226,19 +226,26 @@ func (jr *jReceiver) stopTraceReceptionLocked(ctx context.Context) error {
...
 func (jr *jReceiver) SubmitBatches(ctx thrift.Context, batches []*jaeger.Batch) ([]*jaeger.BatchSubmitResponse, error) {
-	jbsr := make([]*jaeger.BatchSubmitResponse, 0, len(batches))
-	spansMetricsFn := internal.NewReceivedSpansRecorderStreaming(ctx, "jaeger-collector")
+	var err error
+	var nSpans int64
+
+	observabilityRecorder := internal.NewExporterEventRecorder("jaeger-collector")
+
+	octx := observabilityRecorder.Start(ctx, nil)
+	defer observabilityRecorder.End(octx, nSpans, err)
```
Comment (sjkaris) [inline receiver/jaegerreceiver/trace_receiver.go]: same problem with `err` and `nSpans` -- I think this exists everywhere its used, so not going to comment on each place

**Comment by songy23** [comment]
Any updates on this PR?

**Comment by bogdandrutu** [comment]
I did something different to achieve the same result https://github.com/census-instrumentation/opencensus-service/pull/484

**Comment by pjanotti** [comment]
> I did something different to achieve the same result #484

@odeke-em I am closing this one in favor of #484 - let's know if you think it should be otherwise.

**Pull request closed without merging**